proposals/001-operator-design.md

```diff
@@ -0,0 +1,818 @@
+```
+
+#### SNI networking required
+
+A consequence of this Gateway compatibility semantic is that it’s incompatible with the port-per-broker networking in
+the proxy. That’s because port-per-broker necessarily requires the specification of multiple ports, but
+Gateway.spec.listeners[] has a required, but single-valued, port field. That’s OK, because we were expecting to support
+only SNI networking, though that can require more work by the user in terms of supplying all the necessary certificates.
+
+#### Example KafkaGateway status
+
+Applying the compatibility principle we can see that the schema for the KafkaGateway CR’s status must be the same as for
+Gateway’s status, with conditions as well as listener and address specific statuses.
+
+```yaml
+---
+kind: KafkaGateway
+apiVersion: kroxylicious.io/v1alpha1
+metadata:
+  name: simple
+spec:
+# ...
+status:
+  conditions:
+    - observedGeneration: 1 # the KafkaRoute's generation
+      type: Ready
+      status: True
+  listeners:
+    - name: my-listener
+      supportedKinds:
+        - group: kroxylicious.io
+          kind: KafkaRoute
+  addresses:
+    -
+```
+
+### KafkaGatewayParameters Custom Resource
+
+This is the other end of the KafkaGateway.spec.infrastructure.parametersRef reference. It’s essentially a place to dump
+all the proxy config which doesn’t fit nicely into the KafkaGateway (aka, Gateway) API.
+
+#### Example KafkaGatewayParameters CR
+
+```yaml
+---
+kind: KafkaGatewayParameters
+apiVersion: kroxylicious.io/v1alpha1
+metadata:
+  name: my-params
+spec:
+  replicas: 3 # the number of proxy pods to create
+  management: # optional, but omission =/=> no http server, 
+  # because we always expose kube health endpoints
+  port: 8080 # this port is also used to expose kube health
+  prometheusMetrics: true
+```
+
+### KafkaBackendTLSPolicy Custom Resource
+
+We need to support customising TLS trusted certificates on the backend connection to the target cluster.
+
+In the Gateway API, GatewaySpec has a backendTLS property that contains a reference to a single clientCertificateRef for
+that Gateway.
+
+The use of spec.backendTLS on its own limits possibilities by requiring that all Kafka clusters are configured to trust
+a common CA certificate used to sign the clientCertificateRef.
+
+The Gateway API also has a BackendTLSPolicy (at v1alpha3), described here, which could be used to support backends with
+different trusted CA certificates within the same gateway. However, that’s still a bit of a compromise:
+BackendTLSPolicy could be limiting in the long term if proxy was performing its own SASL authentication with the
+brokers (i.e. how a proxy connect to a backend in general might depend on more that hostname, port and TLS settings, but
+also some SASL settings).
+
+```yaml
+# example with caCerts
+---
+kind: KafkaBackendTLSPolicy
+apiVersion: kroxylicious.io/v1alpha1
+metadata:
+  name: my-tls-policy
+spec:
+  targetRefs:
+    - kind: Service
```

Comment:
this assumes the backend is represented by a service.

Reply:
We could use an ExternalName service but it's a bit indirect yes

Reply:
https://kubernetes.io/docs/concepts/services-networking/service/#externalname doesn't include a port.  I think we'd need our reference object.